### ISSUE TITLE
org.eclipse.xtext.xbase.lib.ArrayExtensions.set stores incorrect value when chained (sometimes silently)

### ISSUE THREAD

**Issue by djozis** [eclipse/xtext-lib#60]
Because `org.eclipse.xtext.xbase.lib.ArrayExtensions.set` inlines as `$1[$2] = $3` and never wraps with parentheses, the inline behavior differs from the logical behavior. Example:
```
val a = <String>newArrayOfSize(1)
val int k = a.set(0, "").length
```
Results in:
```
final String[] a = new String[1];
final int k = a[0] = "".length();
```
Instead of:
```
final String[] a = new String[1];
final int k = (a[0] = "").length();
```
This example produces a compile error because an int is not an String, but in scenarios in which the chained method returns the same type as the expression used in set, it silently inserts the wrong value into the array (which is a frustrating way to stumble on this).

**Comment by djozis** [comment]
Also, wrapping the call to `set` in parentheses doesn't help as they seem to get removed.

**Comment by cdietrich** [comment]
@szarnekow do you see any chance to fix this besides not inlineing?

using parenthesis will not help since

```
val a = <String>newArrayOfSize(1)
a.set(0, "").length // (a[0] = "").length(); ok in java
a.set(0,"hello") // (a[0] = "hello"); no ok in java
```

**Comment by djozis** [comment]
Returning void (like most setters do) could be an option in the interest of still inlining.

**Comment by cdietrich** [comment]
@szarnekow any hints? am not sure if the Inline.statementxxxx should be evaluated here and the .length call should be forbidden

**Comment by cdietrich** [comment]
Fixed in 2.18
